enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,193 @@
+
+## Motivation
+
+By design when OLM uninstalls an operator it does not remove any of the operator’s owned CustomResourceDefinitions (CRDs) and CRs in order to prevent data loss.
+The operator user or cluster admin is expected to manually delete CRs and any other secondary resources created by the operator.
+
+This cleanup can be incomplete when the operator is no longer available to process deletion events of CRs and handle cleanup via finalizers.
+Secondary resources that have their [OwnerReference][owner-reference] set to a CR would be garbage collected, but others such as off-cluster or cross-namespaced resources with no owner references would not be removed if the operator is already removed.
+
+So to facilitate automatically removing operator managed resources OLM can trigger cleanup via CR removal before the operator is removed.
+
+### Goals
+
+- Clean up includes removal of CRs (operands) and CRDs
+- CRs and CRDs are not automatically removed
+- A user can opt-in to the clean up behavior at operator uninstall time
+
+### Non-Goals
+
+- OLM handles cleanup of anything other than CRs
+- Handle custom cleanup logic like filtering/ordering the deletion of CRs
+
+### Possible Future Goals (Not in scope)
+
+- OLM should fire an alert event if the operator is stalled or taking too long on operand cleanup
+- The timeout for firing an alert event should be user configurable
+- Once [operator conditions][operator-conditions] are supported, there could be an OLM supported condition that the operator can use to indicate operand cleanup failure
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator user I want my operator uninstall to also clean up operator-created resources so that I don't have to manually clean up any leftover resources.
+
+### Implementation Details/Notes/Constraints
+
+#### Where to configure cleanup: CSV and/or Subscription
+
+The API to configure operand cleanup could be specified on either the Subscription or ClusterServiceVersion (CSV).
+
+If the cleanup API is only present on the Subscription then operators installed without Subscriptions (by manually creating the CSV and fulfilling requirements) can not opt-in to operand cleanup.
+
+The CSV is the more appropriate point to configure the cleanup option for an operator as the CSV represents the actual installation of the operator, whereas the Subscription only represents the intent to subscribe to updates from the catalog. Moreover, since deleting the CSV causes the operator to be removed, deleting the CSV should be the trigger to initiate operand cleanup. See the [section on CSV finalizers][finalizers-on-csvs] for more details.
```

Comment:
https://github.com/operator-framework/enhancements/pull/46#discussion_r511030854
>  I am a little worried this would create the scenario where somehow the cleanup toggles get out of sync between Subscription and CSV. I'd lean towards only Subscription controlling the cleanup behavior as it also controls the update behavior of the CSV.

@dmesser Just wanted to make sure we can address your previous comments here.

Per my [earlier comment](https://github.com/operator-framework/enhancements/pull/46#discussion_r511101669) the CSV seems to be the better place to configure the cleanup option.
Along with the UX of CSV deletion being the trigger point for operand cleanup.

Let me know what you think, since the rest of the proposal on finalizers and showing the operand cleanup status depends on where we place the cleanup option.

Reply:
Thanks @hasbro17. I follow your reasoning about putting it on the `CSV` rather than the `Subscription`.

Reply:
By placing the `cleanup function` on the CSV, users that install an operator via a subscription are required to take an additional step to enable/disable cleanup by editing the generated CSV. In the past, users have relied on the [subscriptionconfig](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/design/subscription-config.md) resource to modify the generated CSV.

I am also concerned that this field is initially set via the Operator Author rather than the user. A new OLM user may not know that some operators allow cleanup while others do not and I am afraid that this jarring experience may hurt their perception of OLM. Ideally, the CSV could convey support for operand deletion as well as the "switch" to enable/disable the feature while the subscription could override the existing field in the CSV.

Reply:
@njhale I agree that Subscription alternative could probably be in the alternatives section. 
However since it's been such a central point of contention/discussion as the proposal evolved from using Subscription to CSV that I felt it made more sense to explain the rationale of using the CSV upfront and contrast the UX/pros/cons with the Subscription.
However if folks still feel it's confusing I can move the Subscription vs CSV debate down into the alternative section.

@awgreene There are tradeoffs in making the CSV the sole configuration point for the cleanup switch.
The thread in https://github.com/operator-framework/enhancements/pull/46#discussion_r497707442 has some context.

> users that install an operator via a subscription are required to take an additional step to enable/disable cleanup by editing the generated CSV

The decision to opt-in/opt-out of cleanup can be made up until operator/CSV removal time. And it's likely that's users will toggle the switch closer to uninstall time which is when they consider the cleanup behavior. So that extra step of updating the CSV (or Subscription) after install would likely still be necessary.

I initially did have the Subscription-CSV pair configuration in mind. Although the Subscription `spec.config` currently is just overriding configurations in the operator deployment/pod, not the CSV.

Having the cleanup config be present on the Subscription that is then synced to and overwritten on the CSV gives the user two points to configure the cleanup toggle.
These toggles can become out of sync if the user installs the operator with the cleanup option enabled on the Subscription `subscription.spec.cleanup: true` but later toggles the field directly on the CSV `csv.spec.cleanup: false`.
Now the subscription is declares one thing while the CSV will effect the opposite.

I can see the merits of the dual Subscription-CSV config but I have some doubts over the UX. I'll let @dmesser chime in on that though.

Reply:
> I am also concerned that this field is initially set via the Operator Author rather than the user

This is a valid concern. One that @joelanford brought up as well in a [previous comment](https://github.com/operator-framework/enhancements/pull/46#discussion_r517648737) above.
An operator user might not realize that a CSV author has by default opted into cleanup with `spec.cleanup: true`, until it's too late and after their operands are cleaned up.

I don't think the "Subscription-overrides-CSV" config approach fixes this problem since it doesn't cover the case of subscription-less installs. And I've already mentioned its other drawback in the comment above.

So I think we need to decide if we're okay with letting CSV authors set the default for the cleanup toggle, potentially catching users who are unaware of the CSV config off guard.
Or if we want the opt-in to be done only by the operator user and explicitly after install time (i.e the CSV is always created with `spec.cleanup: false`).

Haven't thought too much about how this might be done (possibly a validating admission webhook for the CSV) but the arguable drawback here is that it makes for a weird UX if you can't set a CSV field at create time but can do so after creation (not sure if there is a precedence here).

So @dmesser @njhale @awgreene @ecordell I could some input on the UX here when it comes to allowing/disallowing the CSV author to set the default cleanup option.

Reply:
For the sake of simplicity we should retain the current behavior where the control over the user data remains on the user side. This is in order to avoid surprises due to different behavior of Operators getting removed when there was active decision by the user. There should be a consistent default which is "no Operands are deleted upon Operator remove".

We can always revisit this later.

Reply:
I worry that we’re giving the user too much credit here. Do they really know and understand OLM well enough that the precedent of “no cleanup” is an expectation? I can tell you it surprised me initially: I first thought it was something I had missed doing in my operator; and then that it was simply a huge gaping bug in OLM.

It’s also worth noting that cleanup doesn’t necessarily result in data loss. In my case, the CRs bridge to secondary resources, but those aren’t deleted when the CR is removed. And that’s the kind of thing that only the operator author knows. 

So giving the reins to the operator author, and letting them call out the behavior in their operator’s documentation, still has my vote, for what it’s worth.

Reply:
I sort of have to disagree here. Sure, there will be a few cases where the cleanup steps are harmless and the operator developer knows best. But I think we have more than a few examples where the developer is just unaware of what features and settings they are opting into. The reverse case here is catastrophic data loss for a user. I think this is one of the situations where we want to do everything we can to make it as difficult as possible to opt into this feature -- it should be as close to the experience of having to say "are you really really sure?" Every single time you use it.

The actual use case for this feature is, imo, a cluster admin who is administrating large clusters with potentially very complex inputs and outputs and wants the option to be able to automate this. The worst case scenario for the reverse is that that admin needs to perform some additional cleanup steps, or reinstall the operator and flip the toggle before removing it a second time. If that means that a cluster admin who administrates a single cluster doesn't accidentally lose their database, I would strongly prefer a slightly clunky ux

Reply:
>  I would strongly prefer a slightly clunky ux

Maybe we can avoid this by making cleanup a Subscription-only feature.

(I think we've come full circle here)

Reply:
I would rather give the developer credit for knowing/investigating "what features and settings they are opting into" than expecting the same of the user.

That said, I can get behind "safer is better", given that the "reinstall, flip the switch, re-uninstall" option exists. Puts a little extra onus on the developer to make sure reinstall picks up existing resources, but that should be the case anyway.